controllers/halayerset_controller_test.go

```diff
@@ -175,52 +178,104 @@ var _ = Describe("High Availability Layer Set CR", func() {
+		verifyNoUnExpectedCommand(forbiddenCommands)
 	case deleteCR:
-		Eventually(func() int { return len(execCmdCommands) }, time.Second*10, time.Millisecond*10).Should(BeEquivalentTo(3))
 		expectedCommands := [][]string{
 			{"pcs", "status", "xml"},
-			{"pcs", "stonith", "create", originalFenceAgentName, fenceAgentType},
-			{"pcs", "resource", "remove", originalFenceAgentName, "--force"},
+			{"pcs", "stonith", "create", firstOrgFenceAgent, fenceAgentType},
+			{"pcs", "resource", "remove", firstOrgFenceAgent, "--force"},
+		}
+		forbiddenCommands := [][]string{
+			{"pcs", "resource", "remove", updatedFenceAgentName, "--force"},
+			{"pcs", "stonith", "create", updatedFenceAgentName, fenceAgentType},
 		}
 		verifyExpectedCommands(expectedCommands)
+		verifyNoUnExpectedCommand(forbiddenCommands)
 	case updateCR:
-		Eventually(func() int { return len(execCmdCommands) }, time.Second*10, time.Millisecond*10).Should(BeEquivalentTo(5))
 		expectedCommands := [][]string{
-			{"pcs", "stonith", "create", originalFenceAgentName, fenceAgentType},
+			{"pcs", "stonith", "create", firstOrgFenceAgent, fenceAgentType},
+			{"pcs", "stonith", "create", secondOrgFenceAgent, fenceAgentType},
 			{"pcs", "status", "xml"}, //appears twice
-			{"pcs", "resource", "remove", originalFenceAgentName, "--force"},
+			{"pcs", "resource", "remove", firstOrgFenceAgent, "--force"},
 			{"pcs", "stonith", "create", updatedFenceAgentName, fenceAgentType},
+			{"pcs", "stonith", "update", secondOrgFenceAgent},
 		}
+		forbiddenCommands := [][]string{
+			{"pcs", "resource", "remove", updatedFenceAgentName, "--force"},
+			{"pcs", "resource", "remove", secondOrgFenceAgent, "--force"},
+		}
+		//debug()
 		verifyExpectedCommands(expectedCommands)
+		verifyNoUnExpectedCommand(forbiddenCommands)
 	}
 
 }
 
+func debug() {
+	for i := 0; i < 30; i++ {
+		time.Sleep(time.Second)
+	}
+	fmt.Println("Done")
+}
+
 func verifyExpectedCommands(expectedCommands [][]string) {
-	var actualCommands [][]string
+	Eventually(isExpectedCommandsFound(expectedCommands), time.Second*10, time.Millisecond*10).Should(BeTrue())
+}
+
+func isExpectedCommandsFound(expectedCommands [][]string) bool {
+	for _, expectedCommand := range expectedCommands {
+		if !containsStringSlice(expectedCommand, getActualCommands()) {
+			return false
+		}
+	}
+	return true
+}
+
+func getActualCommands() [][]string {
+
 	for len(execCmdCommands) > 0 {
-		actualCommand := <-execCmdCommands
-		actualCommands = append(actualCommands, actualCommand.command)
+		actualCommands = append(actualCommands, <-execCmdCommands)
 	}
-	for _, expectedCommand := range expectedCommands {
-		Expect(containsStringSlice(expectedCommand, actualCommands)).To(BeTrue())
+
+	return actualCommands
+}
+
+func verifyNoUnExpectedCommand(forbiddenCommands [][]string) {
+	for _, forbiddenCommand := range forbiddenCommands {
+		for _, actualCommand := range getActualCommands() {
+			if len(actualCommand) >= len(forbiddenCommand) {
```

Comment:
Why do you check that? If that't not necessary, then we can run this function with two loops in o(n^2) instead of three loops in o(n^3)...

Reply:
- Since this is a unit test code run time is not critical here.
- Note the line you are referring is an `if` statement and not a `for` loop so it's still an o(n^2)
- The reason I use it is because in this case `actualCommand` containing `forbiddenCommand` is enough to fail the test so no need to check for an exact match.